### ISSUE TITLE
Gomod for public API broke controller-tools object generation

### ISSUE THREAD

**Issue by s3rj1k** [metal3-io/baremetal-operator#875]
After https://github.com/metal3-io/baremetal-operator/commit/f4865dbd25f288543513b1afb3f0e614070ad226
There is no simple way on regenerating objects for public API.

Current `make generate` does **not** generate objects for public API (silently ignores).

To actually see that object generation is broken, one would need to run
```
go run sigs.k8s.io/controller-tools/cmd/controller-gen object paths="./apis/..."
```

Output:
```
-: pattern ./apis/...: main module (github.com/metal3-io/baremetal-operator) does not contain package github.com/metal3-io/baremetal-operator/apis/metal3.io/v1alpha1
Error: not all generators ran successfully
run `controller-gen object paths=./apis/... -w` to see all available markers, or `controller-gen object paths=./apis/... -h` for usage
exit status 1

```

To manually regenerate objects, as a temporary measure, one would need to remove gomod related files from `apis` folder and remove related `requires` and `replaces` from repository root go.mod.

**Comment by s3rj1k** [comment]
highlighting everyone who participated in https://github.com/metal3-io/baremetal-operator/pull/857
/cc: @kashifest @dhellmann @zaneb @maelk

**Comment by dhellmann** [comment]
> We also could try to move main.go and go.mod out of root (`cmd` dir maybe?), did not test that.

It's worth experimenting with that, but I expect kubebuilder would have trouble finding the file to add new things to it when we add a controller or API version.

One approach we discussed and rejected before we turned `apis` into a separate module was using a separate git repo. How much would that save us in terms of dependencies, if there was only an API and no matching controller?

**Comment by dhellmann** [comment]
> > One approach we discussed and rejected before we turned apis into a separate module was using a separate git repo.
> 
> You mean like git sub-module? That's definitely harder to manage.
> 
> If we are talking about regular git repo separated from this one, that could work in terms of making dependencies as small as possible but same question for kubebuilder part, will this work?

This second idea is the approach I meant.

We should be able to have 1 kubebuilder repo with just an API in it and another with the controllers, provisioners, etc. Our controller already imports and uses other API definitions like Secret.

> Also there is a question about version matching, repos should start using tags and matching releases with semantic versioning on `minor`.

Yes, tagging the API repo more often would make consuming it easier.

What difficulties would splitting the APIs into a separate repo introduce? Are we likely to regret doing it?

**Comment by zaneb** [comment]
Tangentially related problem: https://github.com/metal3-io/baremetal-operator/pull/865#discussion_r620373011

I _think_ a solution to that is to make it possible to import the api package without automatically registering the type to a GVK. Then in the baremetal-operator we could register a type that inherits from BareMetalHost and adds validation code but exists in a different package. We could have a subpackage or something to do the registration.

Obviously that breaks any existing stuff that is importing the API (e.g. CAPM3). I think my point here is that if we're going to break consumers in another way also, then we should do this at the same time.

**Comment by maelk** [comment]
It should also possible to run the executables in the apis folder. That allows to generate the manifests and others without changes. Anything that we do at the root of the repo should be doable in the apis folder too afaik.

**Comment by maelk** [comment]
That would be exactly the same solution as https://github.com/metal3-io/baremetal-operator/pull/871

**Comment by s3rj1k** [comment]
@maelk Separate binary does not work, this is due to a problem being in separate go.mod for sub package and root repo go.mod.

**Comment by furkatgofurov7** [comment]
We were seeing the same problem when working with putting go modules same way in IPAM and solution was [af0e50e](https://github.com/metal3-io/ip-address-manager/pull/32/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52R172-R176)

**Comment by kashifest** [comment]
Could this be a solution https://github.com/metal3-io/baremetal-operator/pull/877 ?

**Comment by s3rj1k** [comment]
@furkatgofurov7 @kashifest both of the solutions do not work as one would expect

![изображение](https://user-images.githubusercontent.com/11349489/117948333-b6f50f80-b319-11eb-80f2-27e4474f37dc.png)

**Comment by s3rj1k** [comment]
on the other hand, using binaries directly does work as expected

![изображение](https://user-images.githubusercontent.com/11349489/117948957-4dc1cc00-b31a-11eb-94ef-96f2a1183ca9.png)

objects are generated and no new `require` packages are needed

**Comment by s3rj1k** [comment]
(sorry for none-English locale)

**Comment by dtantsur** [comment]
Hint re locale:
```
$ LC_ALL=ru_RU.utf-8 git status
На ветке main
Ваша ветка обновлена в соответствии с «origin/main».

нечего коммитить, нет изменений в рабочем каталоге
```
```
$ LC_ALL=en_US.utf-8 git status
On branch main
Your branch is up to date with 'origin/main'.

nothing to commit, working tree clean
```
(I have a habit of setting the system locale to English specifically to avoid scaring people with cyrillics :D )

**Comment by kashifest** [comment]
> @furkatgofurov7 @kashifest both of the solutions do not work as one would expect
> 
> ![изображение](https://user-images.githubusercontent.com/11349489/117948333-b6f50f80-b319-11eb-80f2-27e4474f37dc.png)

I didn't understand, the PR seems to work no?

**Comment by s3rj1k** [comment]
@dtantsur yea, need to add some alias to .bashrc for this :)))

**Comment by s3rj1k** [comment]
@kashifest it works, yes, but you introduce new unneeded dependency in public API, the whole point of go.mod in public API was to allow external users to import package with minimal deps.

Also you would have to track `two` controller-tools and insure that both of them are of the same version, this is also seems suboptimal

**Comment by kashifest** [comment]
> @kashifest it works, yes, but you introduce new unneeded dependency in public API, the whole point of go.mod in public API was to allow external users to import package with minimal deps.
> 
> Also you would have to track `two` controller-tools and insure that both of them are of the same version, this is also seems suboptimal

Well I agree. Its adding two new dependencies which are not needed by api itself but the generate process. But comparing the complicacy of not using a go.mod in apis and syncing with CAPI/CAPM3 , it is still less. If you have other solutions, its welcome. But before reverting #857 we need to ensure that the new solution keeps CAPM3 imports intact also.

**Comment by s3rj1k** [comment]
@kashifest instead of using `go run sigs.k8s.io/controller-tools/cmd/controller-gen` lets just use binary directly, problem solved

**Comment by s3rj1k** [comment]
https://github.com/metal3-io/baremetal-operator/commit/b3e23f2b487a02f770b886c548f469818d613aab

**Comment by kashifest** [comment]
> @kashifest instead of using `go run sigs.k8s.io/controller-tools/cmd/controller-gen` lets just use binary directly, problem solved

+1

**Comment by s3rj1k** [comment]
cc: @asalkeld

**Comment by ardaguclu** [comment]
What we gain from go.mod separation in https://github.com/metal3-io/baremetal-operator/pull/857;
- CAPM3 version support as clearly stated in PR by @kashifest .
It is a very clear need and I don't anything to say about it. Because you know better than me why we are not synchronizing versions one to an another and why adding go.mod into `apis/` to solve this problem.
- Smaller size of api package
I assume, consumers are more likely to import pkg as well and the actual size is the same for most cases. Correct me if I'm wrong.
- Ability to manage different package versions in these 2 go modules.
Does it the case?, can we now use for example different controller-runtime versions in `apis/` and `pkg/` (currently we are using the same 0.7.0). For now, @s3rj1k's suggestion using controller-gen binary as a unique binary seems working solution but in the future, if controller-runtime or other package versions(I mean major gaps) differ in these 2 modules, will one controller-gen binary solve our problem(one linter binary probably will solve).

But up to now, these are all concerns and go mod separation is still very relevant because of the **CAPM3 version support** need.

But I'd like to add my opinions after spending some time to https://github.com/metal3-io/baremetal-operator/pull/865;
For general webhook feature, we have 3 options;
- Putting validation logic in `apis/`
This creates conflict of some logic will be under `pkg/` and some logic inevitably will be under `apis/`. We have 2 modules but some of the bootmacaddress checks are dispersed between pkg and apis.
- Not introducing any webhook.
I suppose, this is not acceptable.
- Applying @zaneb's idea, moving GVK register code under `pkg/`. This can work but we'll change the `apis/` and what will be left under `apis/` after moving the registration under `pkg/`. We will manage go module for a very small piece of code under `apis/`. I don't know, maybe this is the right thing.

To sum up, maybe instead trying to manage `make` commands, manage different go modules, we can revert back to one go module, after of course finding a way to version conflict which @kashifest stated.

**Comment by zaneb** [comment]
The main reason for the separate go.mod files AIUI is that when you import a package from a module, every dependency in that module's go.mod participates in the dependency solving - even ones that aren't imported by the packages you're actually using. The larger this set is, the more likely it is to result in an unresolvable dependency (due to mismatched version requirements) for somebody somewhere. CAPM3 seems to have hit it hardest, but other people have complained.

**Comment by andfasano** [comment]
After some thoughts I'd agree with @dhellmann for the separate API github repo approach (ie `baremetal-api`)

Even thought #857 appeared as an interesting approach, the management of a multi-module project in a single repo still looks like difficult, especially for those part that could have a different lifecycle in future.
Keeping the two projects separated instead could have some valuable benefits:

* API and implementation could have different versions (and API will tend obviously to be more stable) - separate lifecycle and maintenance
* External customers will refer just to the API module

Of course #865 must be still be kept in the loop, but still looks like an approach worth to be analyzed, since it could address such issues directly from an architectural point of view.

**Comment by asalkeld** [comment]
> on the other hand, using binaries directly does work as expected
> 
> ![изображение](https://user-images.githubusercontent.com/11349489/117948957-4dc1cc00-b31a-11eb-94ef-96f2a1183ca9.png)
> 
> objects are generated and no new `require` packages are needed

Why not just add a tools.go to the apis directory?

**Comment by maelk** [comment]
@ardaguclu I don't think we can move the GVK registration outside of the API module, this is needed for CAPM3 and other consumers to properly register the GVK when interacting with the BMH object, so this solution would not work. I would really support going with webhooks validation and bringing all validation there if possible, that's the path taken in CAPI / CAPM3 etc.

**Comment by zaneb** [comment]
> I don't think we can move the GVK registration outside of the API module, this is needed for CAPM3 and other consumers to properly register the GVK when interacting with the BMH object, so this solution would not work.

So the idea is that we would put it into a subpackage, like `github.com/metal3-io/apis/metal3.io/v1alpha1/register` and then wherever you need to vendor it you do:

```
import (
   metal3v1alpha1 "github.com/metal3-io/apis/metal3.io/v1alpha1"
    _ "github.com/metal3-io/apis/metal3.io/v1alpha1/register"
)
```

Obviously the downside of this is that you have to explicitly put that second import wherever you are vendoring the module. But if we're going to change the location anyway you can do both at the same time.

That would free us up to _not_ do that subpackage import in baremetal-operator, but instead register a derived type with more methods added for the GVK.

The only alternative I'm seeing would be to write a standalone webhook outside of the framework provided by controller-runtime. We need to pick one of these options if we want to do even some fairly basic validations, like checking if the BMC address is properly formatted and corresponds to a valid driver.

**Comment by dhellmann** [comment]
Can someone remind me which dependency caused the problem in the first place? Is that dependency conflict still an issue, or could we go back to 1 module while we explore other options?

**Comment by zaneb** [comment]
> Can someone remind me which dependency caused the problem in the first place?

It's controller-runtime (see #857).

**Comment by s3rj1k** [comment]
@zaneb 
> The only alternative I'm seeing would be to write a standalone webhook outside of the framework provided by controller-runtime. We need to pick one of these options if we want to do even some fairly basic validations, like checking if the BMC address is properly formatted and corresponds to a valid driver.

This is actually a good alternative, one could replace that webhook with custom version when some specific validation is required and it is not possible to contribute that validation code to upstream.

**Comment by ardaguclu** [comment]
> > Can someone remind me which dependency caused the problem in the first place?
> 
> It's controller-runtime (see #857).

But currently controller-runtime versions are same `v0.7.0`. Maybe some other dependency...

**Comment by andfasano** [comment]
> > > Can someone remind me which dependency caused the problem in the first place?
> > 
> > 
> > It's controller-runtime (see #857).
> 
> But currently controller-runtime versions are same `v0.7.0`. Maybe some other dependency...

IINW this one https://github.com/metal3-io/cluster-api-provider-metal3/blob/b516aeab24e3b1c7b56934fb38ded5e9234002f1/go.mod#L53 ?
